pkg/inventory/inventory-client.go

```diff
@@ -79,79 +62,157 @@ func NewInventoryClient(factory util.Factory) (*ClusterInventoryClient, error) {
+		if err := cic.createInventoryObj(invInfo); err != nil {
+			return nil, err
+		}
+	} else {
+		// Update existing cluster inventory with merged union of objects
+		clusterObjs, err := cic.GetClusterObjs(localInv)
+		if err != nil {
+			return pruneIds, err
+		}
+		if object.SetEquals(objs, clusterObjs) {
+			klog.V(4).Infof("applied objects same as cluster inventory: do nothing")
+			return pruneIds, nil
+		}
+		pruneIds = object.SetDiff(clusterObjs, objs)
+		unionObjs := object.Union(clusterObjs, objs)
+		klog.V(4).Infof("num objects to prune: %d", len(pruneIds))
+		klog.V(4).Infof("num merged objects to store in inventory: %d", len(unionObjs))
+		wrappedInv := WrapInventoryObj(clusterInv)
+		if err = wrappedInv.Store(unionObjs); err != nil {
+			return pruneIds, err
+		}
+		if !cic.dryRun {
+			clusterInv, err = wrappedInv.GetObject()
+			if err != nil {
+				return pruneIds, err
+			}
+			klog.V(4).Infof("update cluster inventory: %s/%s", clusterInv.Namespace, clusterInv.Name)
+			if err := cic.applyInventoryObj(clusterInv); err != nil {
+				return pruneIds, err
+			}
+		}
 	}
-	return UnionPastObjs(prevInventories)
+
+	return pruneIds, nil
 }
 
-// GetPreviousInventoryObjects returns the set of inventory objects
-// that have the same label as the current inventory object. Removes
-// the current inventory object from this set. Returns an error
-// if there is a problem retrieving the inventory objects.
-func (cic *ClusterInventoryClient) GetPreviousInventoryObjects(currentInv *resource.Info) ([]*resource.Info, error) {
-	current, err := infoToObjMetadata(currentInv)
+// Replace stores the passed objects in the cluster inventory object, or
+// an error if one occurred.
+func (cic *ClusterInventoryClient) Replace(localInv *resource.Info, objs []object.ObjMetadata) error {
+	clusterObjs, err := cic.GetClusterObjs(localInv)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	label, err := retrieveInventoryLabel(currentInv.Object)
-	if err != nil {
-		return nil, err
+	if object.SetEquals(objs, clusterObjs) {
+		klog.V(4).Infof("applied objects same as cluster inventory: do nothing")
+		return nil
 	}
-	prevInventoryObjs, err := cic.retrievePreviousInventoryObjects(current, label)
+	clusterInv, err := cic.getClusterInventoryInfo(localInv)
 	if err != nil {
-		return nil, err
+		return err
+	}
+	wrappedInv := WrapInventoryObj(clusterInv)
+	if err = wrappedInv.Store(objs); err != nil {
+		return err
 	}
-	// Remove the current inventory info from the previous inventory infos.
-	pastInventoryInfos := []*resource.Info{}
-	for _, pastInfo := range prevInventoryObjs {
-		past, err := infoToObjMetadata(pastInfo)
+	if !cic.dryRun {
+		clusterInv, err = wrappedInv.GetObject()
 		if err != nil {
-			return nil, err
+			return err
 		}
-		if !current.Equals(past) {
-			pastInventoryInfos = append(pastInventoryInfos, pastInfo)
+		klog.V(4).Infof("replace cluster inventory: %s/%s", clusterInv.Namespace, clusterInv.Name)
```

Comment:
So the new single inventory object will have the name namespace/name? What if we extend the functionality to allow objects of different namespaces in an inventory object ? The naming of the inventory object/cm in the cluster is critical.